pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -1195,16 +1214,17 @@ func (m *kubeGenericRuntimeManager) computeInitContainerActions(pod *v1.Pod, pod
 				}
 
 				// If the init container is in unknown state, restart it.
+				message = fmt.Sprintf("Init container is in %q state, try killing it before restart", status.State)
 				changes.ContainersToKill[status.ID] = containerToKillInfo{
 					name:      container.Name,
 					container: container,
-					message: fmt.Sprintf("Init container is in %q state, try killing it before restart",
-						status.State),
-					reason: reasonUnknown,
+					message:   message,
+					reason:    reasonUnknown,
 				}
 				changes.InitContainersToStart = append(changes.InitContainersToStart, i)
 			}
 		}
+		klog.V(4).InfoS("Message for Init Container of pod", "containerName", container.Name, "containerStatusID", status.ID, "pod", klog.KObj(pod), "containerMessage", message)
```

Comment:
is the expectation that the `message` will be initialized above and will not be overridden by other logic? It is very hard to reason about it and check that this will not be broken in future. Can you please elaborate on why you chose this approach?

Reply:
see above

Reply:
The goal was to make clear that if any container needs to be restarted, it should only be restarted once and the only difference in behavior is the reason and message for the restart

As a side note: in this case and in regular containers, there is no containerKillReason specified when the restart is due to a container definition change. It would make sense to add one but that would affect the behavior of regular containers as well.

Reply:
Should this message be located in the `if` above? If `restartContainer` is `false`, the following message that doesn’t look so helpful is emitted in  [kubelet.log](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/126525/pull-kubernetes-node-e2e-containerd/1837262541361778688/artifacts/tmp-node-e2e-8d2c2f8b-cos-beta-117-18613-0-41/kubelet.log):

```
I0920 23:07:59.432179	1509 kuberuntime_container.go:1217] "Message for Init Container of pod" containerName="restartable-init-1" containerStatusID={"Type":"containerd","ID":"6255c74259e41c4dc4a480e2b4b7fc3e020e93d49ec9b34b7a5b3d86acc5e2fc"} pod="containers-lifecycle-test-405/test-pod" containerMessage="Init container restartable-init-1"
```